### PR TITLE
Add property State in additional to ProvisioningState

### PR DESCRIPTION
**Pull request by shihying**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [ ] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@shihying__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (shic). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by azurecla** [comment]
Hi __@shihying__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (shic). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by fearthecowboy** [comment]
Is this a error in the original?

If you regenerate this, it will be a breaking change in the SDK.

**Comment by shihying** [comment]
This is not a error in the original. 
We introduce new property "State" and we are intended to use this "State" for all non-Arm party.

We also have PR in sdk, so breaking change should be covered.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: This is a standard property as per ARM convention. Did you get an approval by the ARM Review board? This looks incorrect to me.
/cc @ravbhatnagar - Can you please take a look at this issue?

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: Yes but we introduced one more "state" property in our ARM response to be used by service other than ARM
The ASAzure side already checked-in related change. There were some review process in service side so that we checked in the change

To be clear, "provisioningState" is still there, but we do not want it get used by SDK.

Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: > Yes but we introduced one more "state" property in our ARM response to be used by service other than ARM. ProvisioningState is still there to be used by ARM.

Why create new things? Why can't the same property be reused to provide the same information?

> To be clear, "provisioningState" is still there, but we do not want it get used by SDK.

Why? The swagger spec should accurately describe the REST API. The spec is used for multiple purpose. It is used for generating REST API documentation (docs.microsoft.com) apart from generating the sdks. So describing the swagger spec accurately is super important.


I am still not convinced why is this change really required and why was the provisioningState not reused. I am sure there might be some valid reason behind this. Can you please explain that in detail so folks in the SDK team can get more context on this?

Comment (ravbhatnagar) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @shihying - Can you please elaborate on how you intend to use the proposed "state" property? Which state is it going to indicate and what are the values that will be returned for this? From above comment it looks like it will be the same (and correct me if I am wrong on this) as "provisioningState" values. If that is the case, please let us know why you would like to add this additional property.

Comment (ravbhatnagar) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @shihying - Can you please elaborate on how you intend to use the proposed "state" property? Which state is it going to indicate and what are the values that will be returned for this? From above comment it looks like it will be the same (and correct me if I am wrong on this) as "provisioningState" values. If that is the case, please let us know why you would like to add this additional property.

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @amarzavery 
I made a wrong change. We want to keep provisioningState in swagger and add new "state" property to let it used by SDK. So there won't be breaking change and REST API is accurately described as it is
@amarzavery @ravbhatnagar 
I would update reason why we need addition state field instead of reuse provisioningState soon

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @amarzavery @ravbhatnagar 

Here is the explanation from our team member:
I think it's clear for the reason why we need state. I would update 2nd iteration to surface both provisioningState and state, thanks 

We will use ‘State’ property to return custom terminal values for resource state.
ARM spec allows ‘Succeeded, Failed, Canceled’ values for resource terminal states.

We want to have custom values for our resources. E.g. AS server can  be in ‘Paused’ terminal state.
Response in that case is properties.State=’Paused’, properties.ProvisioningState=’Succeeded’.
Change to make State and ProvisioningState have different values will be made later.

**Comment by shihying** [comment]
@amarzavery @ravbhatnagar
Please check my latest reply in the thread, a new iteration is pushed.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (athipp) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: "...It allows more custom terminal state like Paused." Terminal states are already covered by ProvisioningState. so change the description to be something like "the state variable is used to indicate the current state of the given Analysis Services server."

Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @athipp - Did you verify with ARM team, whether you can add more values to provisioningState. My understanding is that ARM does not stop an RP to add more values. If that is true then don't you think using one property would be easier. It will be less confusing for customers.
@ravbhatnagar - Can you assert my understanding about ARM in the above statement?

Comment (ravbhatnagar) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @athipp - Thanks for the explanation. One thing to clarify is that RPs can put any number of transient/non-terminal states inside "provisioningState" property. Yes, there is a fixed set of terminal states as you mentioned above. W.r.t the new proposed property, I wanted to confirm one thing-

It looks like the resource will go into these new states outside of resource provisioning. Can you confirm? In this case, we are fine with the new property. However, please consider renaming it to something else like a "status" so that its not confused with provisioningState.

But if it was part of resource provisioning, there is a case for merging it with "provisioningState" and we can discuss this more in person.

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: Yes, state like paused would be the state outside of resource provisioning.
In this case, ProvisioningState='Succeeded', State='Paused'

@ravbhatnagar 
I think it's better to be consistent with name from response in ARM-AnalysisService
Can we keep state unless service side change state to status?

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @amarzavery @ravbhatnagar
The change from service side is after discussion with ARM architect, who has the final say on this.

'State' would be used for the state outside of resource provisioning
We would insist on 'State' as the ARM architect proposed it and renaming in service side would break ASPaaS.

Please let us know if any of you has further comment, thanks

**Comment by shihying** [comment]
@amarzavery @ravbhatnagar
Please check my latest reply in the thread, a new iteration is pushed.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (ravbhatnagar) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @shihying - In that case, lets make sure the description of these two properties in clear in how it can be used. Also, lets have these as enums with the possible values listed for both "state" and "provisioningState". Such things help generate better docs from the swagger and easier to use SDKs.

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @ravbhatnagar 
Sure, would make it clear in description.
For enums :  provisioningState and state would be same until service side made further change.

Should we make it enum after service side make them have different values?

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @ravbhatnagar
I suggest to add enums after service side make further change to differentiate provisioningState and state.
If you still want to have enums for now, I would add same value set for both of them.

Please let us know update.

Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @shihying - I would suggest providing enum values `"enum": ["array of valid values"]` and add the extension `"x-ms-enum": { "name": "State", "modelAsString": true }`. When you set it as modelAsString: true, Autorest will not treat it as an enum. Instead it will be a static class with constant properties. This still provides intellisense to the customer, but there is no validation.  Please do this for both "provisioningState" and "state".

Comment (athipp) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @shihying, the suggestion from @amarzavery is useful and there are examples of it already in the json. Let me know if you need help with it.

Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: @amarzavery @ravbhatnagar @athipp 
I pushed a new iteration which adds enums/x-ms-enum.
Noted that value set would be same at this moment to be consistent with response.
We would change value set to be different in future after ASPAAS team implement it.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: fix indentation.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (amarzavery) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: fix indentation.

**Review by amarzavery — CHANGES_REQUESTED** [review]

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: already fixed in latest iteration. please check.

File: arm-analysisservices/2016-05-16/swagger/analysisservices.json
Comment (shihying) [inline arm-analysisservices/2016-05-16/swagger/analysisservices.json]: already fixed in latest iteration. please check.

**Review by amarzavery — APPROVED** [review]

**Pull request merged**
